fabric-network/src/impl/query/roundrobinqueryhandler.js

```diff
@@ -13,52 +13,46 @@ const util = require('util');
 const logger = require('fabric-network/lib/logger').getLogger('RoundRobinQueryHandler');
 
 class RoundRobinQueryHandler {
-	constructor(peers, options = {}) {
+	constructor(peers) {
 		logger.debug('constructor: peers=%j', peers.map((peer) => peer.name));
 		this._peers = peers;
 		this._currentPeerIndex = 0;
-		this._options = options;
 	}
 
 	async evaluate(query) {
 		const method = 'evaluate';
 		logger.debug('%s - start', method);
 
-		const startPeerIndex = this._currentPeerIndex;
-		const errorMessages = [];
-		const options = {requestTimeout: 3000};
-		// use the timeout as the requestTimeout or let default
-		if (Number.isInteger(this._options.timeout)) {
-			options.requestTimeout = this._options.timeout * 1000; // in ms;
+		if (this._currentPeerIndex >= this._peers.length) {
+			// start over
+			this._currentPeerIndex = 0;
 		}
 
+		const startPeerIndex = this._currentPeerIndex++;
+		const errorMessages = [];
+
 		for (let i = 0; i < this._peers.length; i++) {
 			const peerIndex = (startPeerIndex + i) % this._peers.length;
-			this._currentPeerIndex = peerIndex;
-			const peer = this._peers[peerIndex];
-
-			logger.debug('%s - query sending to peer %s', method, peer.name);
-			const results = await query.send({targets:[peer]}, options);
 
-			if (results.errors.length > 0) {
-				logger.error('%s - problem with query to peer %s error:%s', method, peer.name, results.errors[0]);
-				// since only one peer, only one error
-				errorMessages.push(results.errors[0].message);
-				continue;
-			}
-
-			const endorsementResponse = results.responses[0];
+			const peer = this._peers[peerIndex];
+			logger.debug('%s - sending to peer %s', method, peer.name);
 
-			if (!endorsementResponse.endorsement) {
-				logger.debug('%s - peer response status: %s message: %s',
-					method,
-					endorsementResponse.response.status,
-					endorsementResponse.response.message);
-				throw new Error(endorsementResponse.response.message);
+			const results = await query.evaluate([peer]);
+			const result = results[peer.name];
+			if (result instanceof Error) {
+				errorMessages.push(result.toString());
+			} else {
+				if (result.status < 400) {
```

Comment:
You suggested a better approach was to look for an endorsement being present rather than use the status code, so that is what we are doing in `Transaction.submit()` now. This should probably be consistent.

Reply:
So what I am thinking today is the query is completely different, where not getting an endorsement may be the answer they are looking for, not an error. I think that the handler should be presented with:
status : number
payload : bytes
message : string
isEndorsed : boolean
This should provide all the information required to make a decision to return results or try again with a different set of peers.

Reply:
They are certainly the only things I have used from the proposal response in a query handler implementation. Is there any chance that some custom or future implementation might want to look at anything else contained in the proposal response? If not, this looks like a useful abstraction as the proposal response protobuf isn't the most obvious structure for client code to navigate.

Regardless of how the information is presented to the handler, this refactored handler should continue to behave the same as the one in the 1.4 Node SDK and in the 2.0 Java SDK.

Reply:
The high level "Query" class will process the low level query results and present to the handlers {
status : number
payload : bytes
message : string
isEndorsed : boolean
} keyed by peer name for all peers queried.
The refactored handlers will look at the "isEndorsed" as the indicator to return the payload, this will be behave as they did in 1.4 where only status 200 was returned. The refactored handlers will look at all other non error responses as an error to be thrown as they did in 1.4. The refactored handlers will try another peer if an error is the response, which means a gRPC error, as they did in 1.4. "status" is not used by the included handlers, but may be of values to custom handlers.